janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/strategy/AdjacentVertexHasUniquePropertyOptimizerStrategy.java

```diff
@@ -0,0 +1,145 @@
+import org.apache.tinkerpop.gremlin.process.traversal.dsl.graph.DefaultGraphTraversal;
+import org.apache.tinkerpop.gremlin.process.traversal.step.filter.FilterStep;
+import org.apache.tinkerpop.gremlin.process.traversal.step.filter.HasStep;
+import org.apache.tinkerpop.gremlin.process.traversal.step.map.GraphStep;
+import org.apache.tinkerpop.gremlin.process.traversal.step.map.IdStep;
+import org.apache.tinkerpop.gremlin.process.traversal.step.util.HasContainer;
+import org.apache.tinkerpop.gremlin.process.traversal.util.TraversalHelper;
+import org.apache.tinkerpop.gremlin.structure.Edge;
+import org.apache.tinkerpop.gremlin.structure.Graph;
+import org.apache.tinkerpop.gremlin.structure.Vertex;
+import org.janusgraph.core.Cardinality;
+import org.janusgraph.core.JanusGraphElement;
+import org.janusgraph.graphdb.database.StandardJanusGraph;
+import org.janusgraph.graphdb.query.JanusGraphPredicateUtils;
+import org.janusgraph.graphdb.query.QueryUtil;
+import org.janusgraph.graphdb.query.condition.*;
+import org.janusgraph.graphdb.query.index.IndexSelectionUtil;
+import org.janusgraph.graphdb.tinkerpop.optimize.JanusGraphTraversalUtil;
+import org.janusgraph.graphdb.transaction.StandardJanusGraphTx;
+import org.janusgraph.graphdb.types.CompositeIndexType;
+import org.janusgraph.graphdb.types.IndexType;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static org.janusgraph.graphdb.types.system.ImplicitKey.ADJACENT_ID;
+
+/**
+ * @author Florian Grieskamp (Florian.Grieskamp@gdata.de)
+ */
+public class AdjacentVertexHasUniquePropertyOptimizerStrategy
+    extends AdjacentVertexOptimizerStrategy<HasStep<?>> {
+
+    private static final AdjacentVertexHasUniquePropertyOptimizerStrategy INSTANCE =
+        new AdjacentVertexHasUniquePropertyOptimizerStrategy();
+
+    private AdjacentVertexHasUniquePropertyOptimizerStrategy() {}
+
+    public static AdjacentVertexHasUniquePropertyOptimizerStrategy instance() { return INSTANCE; }
+
+    @Override
+    public Set<Class<? extends ProviderOptimizationStrategy>> applyPost() {
+        Set<Class<? extends ProviderOptimizationStrategy>> postStrategies = new HashSet<>();
+        postStrategies.add(JanusGraphStepStrategy.class);
+        postStrategies.add(JanusGraphLocalQueryOptimizerStrategy.class);
+        return postStrategies;
+    }
+
+    @Override
+    public void apply(final Traversal.Admin<?, ?> traversal) {
+        if (TraversalHelper.onGraphComputer(traversal)
+            || traversal.getStartStep() instanceof TraversalVertexProgramStep // not caught by TraversalHelper
+            || !traversal.getGraph().isPresent()) {
+            return;
+        }
+
+        Graph graph = traversal.getGraph().get();
+        StandardJanusGraph janusGraph = graph instanceof StandardJanusGraph
+            ? (StandardJanusGraph) graph
+            : ((StandardJanusGraphTx) graph).getGraph();
```

Comment:
Curious: when is it a StandardJanusGraph instance and when is it a StandardJanusGraphTx instance?

Reply:
I found out that after introducing the config option, some tests failed because the graph was not a `StandardJanusGraph` as I expected. Instead, the traversal's graph was a `StandardJanusGraphTx`. I did not take a deeper look into the reason for that tbh.

Reply:
Can you point out some test cases where the graph is a StandardJanusGraphTx?

Reply:
`InMemoryJanusGraphComputerTest#testMessagePassingOut` for example. But there are some others in the same test class.
You can just comment out the distinction between `StandardJanusGraph` and `StandardJanusGraphTx` and you will end up with `ClassCastException`s.

Reply:
I can see that comes from FulgoraGraphComputer which I am unfamiliar with. Anyway, there are only two implementations of graph interface in JanusGraph codebase, namely StandardJanusGraph and StandardJanusGraphTx, so I think this is good enough.